alts/shared/uploaders/pulp.py

```diff
@@ -0,0 +1,238 @@
+        reference : str
+            Upload reference in Pulp.
+
+        Returns
+        -------
+        str
+            Reference to the created resource.
+
+        """
+        file_sha256 = hash_file(file_path, hash_type='sha256')
+        response = self._uploads_client.commit(
+            reference, {'sha256': file_sha256})
+        task_result = self._wait_for_task_completion(response.task)
+        return task_result.created_resources[0]
+
+    def _put_large_file(self, file_path: str, reference: str):
+        temp_dir = tempfile.mkdtemp(prefix='pulp_uploader_')
+        try:
+            lower_bytes_limit = 0
+            total_size = os.path.getsize(file_path)
+            self._file_splitter.split(
+                file_path, self._chunk_size, output_dir=temp_dir)
+            for file_ in sorted(os.listdir(temp_dir)):
+                if file_ == 'fs_manifest.csv':
+                    continue
+                split_file_path = os.path.join(temp_dir, file_)
+                # File part may have size lower than self._chunk_size,
+                # so to avoid this issue calculate file size before upload
+                new_size = os.path.getsize(split_file_path)
+                upper_bytes_limit = lower_bytes_limit + new_size - 1
+                self._uploads_client.update(
+                    f'bytes {lower_bytes_limit}-{upper_bytes_limit}/'
+                    f'{total_size}',
+                    reference, split_file_path)
+                lower_bytes_limit += new_size
+        finally:
+            if temp_dir and os.path.exists(temp_dir):
+                shutil.rmtree(temp_dir)
+
+    def _send_file(self, file_path: str):
+        reference, file_size = self._create_upload(file_path)
+        if file_size > self._chunk_size:
+            self._put_large_file(file_path, reference)
+        else:
+            self._uploads_client.update(
+                f'bytes 0-{file_size - 1}/{file_size}',
+                reference,
+                file_path
+            )
+        artifact_href = self._commit_upload(file_path, reference)
+        return artifact_href
+
+    def check_if_artifact_exists(self, sha256: str) -> str:
+        response = self._artifacts_client.list(sha256=sha256)
+        if response.results:
+            return response.results[0].pulp_href
+
+    def upload(self, artifacts_dir: str, **kwargs) -> List[dict]:
+        """
+
+        Parameters
+        ----------
+        artifacts_dir : str
+            Path to files that need to be uploaded.
+
+        Returns
+        -------
+        list
+            List of the references to the artifacts inside Pulp
+
+        """
+        artifacts = []
+        errored_uploads = []
+        for artifact in self.get_artifacts_list(artifacts_dir):
+            try:
+                artifacts.append(self.upload_single_file(artifact))
+            except Exception as e:
+                self._logger.error(f'Cannot upload {artifact}, error: {e}',
```

Comment:
Do not do that anymore)
```
f'Cannot upload {artifact}, error: {e}'
```
better
```
logger.info('Some string %s is got', string)
```
Also, use logger.exception if you want to catch exception info)

Reply:
Fixed